fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/MetaDataEvolutionValidator.java

```diff
@@ -0,0 +1,906 @@
+
+        // Validate that every field in the old descriptor is still in the new descriptor
+        for (Map.Entry<Integer, FieldDescriptor> oldFieldEntry : oldFieldMap.entrySet()) {
+            int fieldNumber = oldFieldEntry.getKey();
+            FieldDescriptor oldField = oldFieldEntry.getValue();
+            FieldDescriptor newField = newFieldMap.get(fieldNumber);
+            if (newField != null) {
+                validateField(oldField, newField, seenDescriptors);
+            } else {
+                throw new MetaDataException("field removed from message descriptor", LogMessageKeys.FIELD_NAME, oldField.getName());
+            }
+        }
+        // Validate that none of the new fields are "required", which would make reading back any existing record impossible
+        for (Map.Entry<Integer, FieldDescriptor> newNumberToDescriptorEntry : newFieldMap.entrySet()) {
+            if (!oldFieldMap.containsKey(newNumberToDescriptorEntry.getKey())) {
+                final FieldDescriptor newFieldDescriptor = newNumberToDescriptorEntry.getValue();
+                if (newFieldDescriptor.isRequired()) {
+                    throw new MetaDataException("required field added to record type", LogMessageKeys.FIELD_NAME, newFieldDescriptor.getName());
+                }
+            }
+        }
+    }
+
+    @Nonnull
+    private Map<Integer, FieldDescriptor> getFieldNumberToDescriptorMap(@Nonnull Descriptor descriptor) {
+        final List<FieldDescriptor> fieldDescriptors = descriptor.getFields();
+        if (fieldDescriptors.isEmpty()) {
+            return Collections.emptyMap();
+        }
+        final Map<Integer, FieldDescriptor> numberToDescriptorMap = new HashMap<>(fieldDescriptors.size() * 2);
+        for (FieldDescriptor fieldDescriptor : fieldDescriptors) {
+            numberToDescriptorMap.put(fieldDescriptor.getNumber(), fieldDescriptor);
+        }
+        return numberToDescriptorMap;
+    }
+
+    private void validateField(@Nonnull FieldDescriptor oldFieldDescriptor, @Nonnull FieldDescriptor newFieldDescriptor,
+                               @Nonnull Set<Pair<Descriptor, Descriptor>> seenDescriptors) {
+        if (!oldFieldDescriptor.getName().equals(newFieldDescriptor.getName())) {
+            // TODO: Field renaming should be allowed with some caveats about if the field is indexed or not
+            throw new MetaDataException("field renamed",
+                    LogMessageKeys.OLD_FIELD_NAME, oldFieldDescriptor.getName(),
+                    LogMessageKeys.NEW_FIELD_NAME, newFieldDescriptor.getName());
+        }
+        if (!oldFieldDescriptor.getType().equals(newFieldDescriptor.getType())) {
+            validateTypeChange(oldFieldDescriptor, newFieldDescriptor);
+        }
+        if (oldFieldDescriptor.isRequired() && !newFieldDescriptor.isRequired()) {
+            throw new MetaDataException("required field is no longer required",
```

Comment:
Isn't required -> optional okay?

Reply:
Hm, I think it might be actually. (I was under the impression it used a different serialization format for required fields when I wrote this, but that's not the case.) The protobuf guidelines seem to warn against downgrading from required to optional. I think the problem case is if the user changes a field from `required` to `optional` and then someone else reads it, they will throw a validation error at deserialization time, which is what the protobuf language guide warns against: https://developers.google.com/protocol-buffers/docs/proto

I guess I mainly included this so that we conformed with the protobuf guidelines on schema evolution (or rather enforced a strict superset of their rules). I suppose we could relax this.

Reply:
Yeah, okay, that sounds right. So I guess we leave it for now.